File: storio-common-annotations-processor/src/test/java/com/pushtorefresh/storio/common/annotations/processor/introspection/StorIOColumnMetaTest.java
package com.pushtorefresh.storio.common.annotations.processor.introspection;

import org.junit.Before;
import org.junit.Test;

import java.lang.annotation.Annotation;

import javax.lang.model.element.Element;

import nl.jqno.equalsverifier.EqualsVerifier;
import nl.jqno.equalsverifier.Warning;

import static org.assertj.core.api.Assertions.assertThat;
import static org.powermock.api.mockito.PowerMockito.mock;
import static org.powermock.reflect.Whitebox.getInternalState;

public class StorIOColumnMetaTest {

	private Annotation annotationMock;
	private Element elementMock;
	private JavaType javaType;

	@Before
	public void setUp() throws Exception {
		annotationMock = mock(Annotation.class);
		elementMock = mock(Element.class);
		javaType = JavaType.BOOLEAN;
	}

	@SuppressWarnings({ "unchecked", "rawtypes" })
	@Test
	public final void constructor() {
		// when
		StorIOColumnMeta storioColumnMeta = new StorIOColumnMeta(elementMock, elementMock, "TEST", javaType,
				annotationMock);

		// then
		assertThat(elementMock).as("Constructor must be set enclosingElement field.")
				.isEqualTo(getInternalState(storioColumnMeta, "enclosingElement"));
		assertThat(elementMock).as("Constructor must be set element field.")
				.isEqualTo(getInternalState(storioColumnMeta, "element"));
		assertThat("TEST").as("Constructor must be set elementName field.")
				.isEqualTo(getInternalState(storioColumnMeta, "elementName"));
		assertThat(javaType).as("Constructor must be set javaType field.")
				.isEqualTo(getInternalState(storioColumnMeta, "javaType"));
		assertThat(annotationMock).as("Constructor must be set storIOColumn field.")
				.isEqualTo(getInternalState(storioColumnMeta, "storIOColumn"));
	}

	@Test
	public final void equalsAndHashCode() {
		EqualsVerifier.forClass(StorIOColumnMeta.class).suppress(Warning.REFERENCE_EQUALITY).usingGetClass().verify();
	}

	@SuppressWarnings({ "unchecked", "rawtypes" })
	@Test
	public final void toStringValidation() {
		// given
		StorIOColumnMeta storioColumnMeta = new StorIOColumnMeta(elementMock, elementMock, "TEST", javaType,
				annotationMock);
		String expectedString = "StorIOColumnMeta{enclosingElement=" + elementMock + ", element=" + elementMock
				+ ", elementName='TEST" + '\'' + ", javaType=" + javaType + ", storIOColumn=" + annotationMock +  '}';

		// when
		String toString = storioColumnMeta.toString();

		// then
		assertThat(expectedString).as("toString method should be equal with expectedString.").isEqualTo(toString);
	}

}

File: storio-common-annotations-processor/src/test/kotlin/com/pushtorefresh/storio/common/annotations/processor/introspection/StorIOColumnMetaTest.kt
package com.pushtorefresh.storio.common.annotations.processor.introspection

import com.nhaarman.mockito_kotlin.mock
import nl.jqno.equalsverifier.EqualsVerifier
import nl.jqno.equalsverifier.Warning
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import javax.lang.model.element.Element

class StorIOColumnMetaTest {

    private lateinit var annotationMock: Annotation
    private lateinit var elementMock: Element
    private lateinit var javaType: JavaType

    @Before
    fun setUp() {
        annotationMock = mock()
        elementMock = mock()
        javaType = JavaType.BOOLEAN
    }

    @Test
    fun constructor() {
        // when
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "TEST", javaType, annotationMock)

        // then
        assertThat(columnMeta.enclosingElement).isEqualTo(elementMock)
        assertThat(columnMeta.element).isEqualTo(elementMock)
        assertThat(columnMeta.elementName).isEqualTo("TEST")
        assertThat(columnMeta.javaType).isEqualTo(javaType)
        assertThat(columnMeta.storIOColumn).isEqualTo(annotationMock)
    }

    @Test
    fun equalsAndHashCode() {
        EqualsVerifier.forClass(StorIOColumnMeta::class.java)
                .suppress(Warning.REFERENCE_EQUALITY)
                .usingGetClass()
                .verify()
    }

    @Test
    fun toStringValidation() {
        // given
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "TEST", javaType, annotationMock, "getter", "setter")
        val expectedString = "StorIOColumnMeta(enclosingElement=$elementMock, element=$elementMock, elementName='TEST', javaType=$javaType, storIOColumn=$annotationMock, getter='getter', setter='setter')"

        // when
        val toString = columnMeta.toString()

        // then
        assertThat(expectedString).isEqualTo(toString)
    }

    @Test
    fun shouldReturnRealElementNameForElementWithoutPrefixes() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "property", javaType, annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("property")
    }

    @Test
    fun shouldReturnRealElementNameForElementWithGetPrefix() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "getProperty", javaType, annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("property")
    }

    @Test
    fun shouldReturnRealElementNameForElementWithIsPrefix() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "isProperty", javaType, annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("property")
    }

    @Test
    fun shouldReturnRealElementNameForElementWithOneCharacterName() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "a", javaType, annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("a")
    }

    @Test
    fun shouldReturnRealElementNameForElementStartsWithGet() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "getter", javaType, annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("getter")
    }

    @Test
    fun shouldReturnRealElementNameForElementStartsWithIs() {
        val columnMeta = StorIOColumnMeta(elementMock, elementMock, "iso", javaType,
                annotationMock)

        val realName = columnMeta.realElementName

        assertThat(realName).isEqualTo("iso")
    }

}
